src/graph/graphcontext.c

```diff
@@ -0,0 +1,196 @@
+#include "graphcontext_type.h"
+#include "../util/rmalloc.h"
+
+// Needed for the key-value creation logic for the Graph
+extern RedisModuleType *GraphRedisModuleType;
+
+int _GraphContext_IndexOffset(GraphContext *gc, const char *label, const char *property) {
+  Index *idx;
+  for (int i = 0; i < gc->index_count; i ++) {
+    idx = gc->indices[i];
+    if (!strcmp(label, idx->label) && !strcmp(property, idx->property)) return i;
+  }
+  return -1;
+}
+
+RedisModuleKey* GraphContext_Key(RedisModuleCtx *ctx, const char *graph_name) {
+  int keylen = strlen(graph_name) + 4;
+  char strKey[keylen + 1];
+  sprintf(strKey, "%s_ctx", graph_name);
+
+  RedisModuleString *rm_graphctx = RedisModule_CreateString(ctx, strKey, keylen);
+  RedisModuleKey *key = RedisModule_OpenKey(ctx, rm_graphctx, REDISMODULE_WRITE);
+  RedisModule_FreeString(ctx, rm_graphctx);
+
+  return key;
+}
+
+GraphContext* GraphContext_New(RedisModuleCtx *ctx, RedisModuleString *rs_name) {
+  // Create key for GraphContext
+  const char *graph_name = RedisModule_StringPtrLen(rs_name, NULL);
+  RedisModuleKey *key = GraphContext_Key(ctx, graph_name);
+  if (RedisModule_ModuleTypeGetType(key) != REDISMODULE_KEYTYPE_EMPTY) {
+    // Return NULL if key already exists
+    RedisModule_CloseKey(key);
+    return NULL;
+  }
+
+  GraphContext *gc = rm_malloc(sizeof(GraphContext));
+  gc->relation_cap = GRAPH_DEFAULT_RELATION_CAP;
+  gc->relation_count = 0;
+  gc->label_cap = GRAPH_DEFAULT_LABEL_CAP;
+  gc->label_count = 0;
+  gc->index_cap = 4;
+  gc->index_count = 0;
+
+  gc->graph_name = rm_strdup(graph_name);
+  gc->node_stores = rm_malloc(gc->label_cap * sizeof(LabelStore*));
+  gc->relation_stores = rm_malloc(gc->relation_cap * sizeof(LabelStore*));
+  gc->indices = rm_malloc(gc->index_cap * sizeof(Index*));
+
+  gc->node_allstore = LabelStore_New("ALL", GRAPH_NO_LABEL);
+  gc->relation_allstore = LabelStore_New("ALL", GRAPH_NO_RELATION);
+
+  // Set and close GraphContext in keyspace
+  RedisModule_ModuleTypeSetValue(key, GraphContextRedisModuleType, gc);
+  RedisModule_CloseKey(key);
+
+  gc->g = Graph_New(GRAPH_DEFAULT_NODE_CAP);
+  // Add new Graph to keyspace
+  // TODO this separation feels rather clunky
+  key = RedisModule_OpenKey(ctx, rs_name, REDISMODULE_WRITE);
+  assert(RedisModule_KeyType(key) == REDISMODULE_KEYTYPE_EMPTY);
+  RedisModule_ModuleTypeSetValue(key, GraphRedisModuleType, gc->g);
+  RedisModule_CloseKey(key);
+
+  return gc;
+}
+
+GraphContext* GraphContext_Get(RedisModuleCtx *ctx, RedisModuleString *rs_name) {
+  const char *graph_name = RedisModule_StringPtrLen(rs_name, NULL);
+  RedisModuleKey *key = GraphContext_Key(ctx, graph_name);
+  if (RedisModule_ModuleTypeGetType(key) != GraphContextRedisModuleType) return NULL;
+
+  GraphContext *gc = RedisModule_ModuleTypeGetValue(key);
+  RedisModule_CloseKey(key);
+
+  gc->g = Graph_Get(ctx, rs_name);
+
+  return gc;
+}
+
+// TODO next two functions are nearly identical; maybe consolidate
```

Comment:
I did write a consolidated version of this, but it seemed unwieldy. I'm currently inclined to leave them as two separate functions (like `Graph_AddLabel` and `Graph_AddRelationType`) - let me know if you have thoughts.